File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution
 {
     public List<Integer> preorderTraversal(TreeNode root)
     {
         var treeVal = new ArrayList<Integer>();
 
         if (root == null) return treeVal;
 
         var tree = new ArrayList<TreeNode>();
         tree.add(root);
 
         while (!tree.isEmpty())
         {
             var node = tree.get(tree.size() - 1);
             tree.remove(tree.size() - 1);
             treeVal.add(node.val);
 
             if (node.right != null)
                 tree.add(node.right);
 
             if (node.left != null)
                 tree.add(node.left);
         }
 
         return treeVal;
     }
 }

File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_3.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var preorderTraversal = function(root) 
{
    const treeVal = []

    if (!root) return treeVal

    const tree = [root]

    while (tree.length > 0)
    {
        const node = tree.pop()
        treeVal.push(node.val)

        if (node.right)
            tree.push(node.right)
        
        if (node.left)
            tree.push(node.left)
    }

    return treeVal
};